### DISCUSSION TITLE
how to setup/configure TFT_eSPI with esp-idf ?

### DISCUSSION THREAD

**Discussion by goutamreddy** [Bodmer/TFT_eSPI#2138] [Q&A - General]
Hi-  I have a basic question about how to use it with esp-idf.
**TL;DR:   Are we supposed to set the TFT settings in menuconfig/sdkconfig   or in the User Setup Select file  or both?**

BACKGROUND
I'm using a LilyGo T-Display.   When I was using Arduino IDE,  I was able to make the library work by changing the following file:
 ~/Documents/Arduino/libraries/TFT_eSPI/User_Setup_Select.h    
      - uncommented `#include <User_Setups/Setup25_TTGO_T_Display.h>`
      - commented  all other  `#include <User_Setup*>` files

I'm now trying to use TFT_eSPI with esp-idf.
I git cloned this TFT_eSPI directory into my esp-idf project component directory (~/ESPprojects/hello_world/components/)
I edited ~/ESPprojects/hello_world/components/TFT_eSPI/User_Setup_Select.h  as above.
Then, I ran idf.py reconfigure....  

When I run idf.py build, I receive a bunch of errors about things not being defined in TFT_config.h:

> /Users/xxx/ESPprojects/hello_world/components/TFT_eSPI/TFT_config.h:128:6: error: #error "Invalid Chip Select pin. Check TFT_eSPI configuration"
>      #error "Invalid Chip Select pin. Check TFT_eSPI configuration"
>       ^~~~~
> /Users/xxx/ESPprojects/hello_world/components/TFT_eSPI/TFT_config.h:134:6: error: #error "Invalid Data/Command pin. Check TFT_eSPI configuration"
>      #error "Invalid Data/Command pin. Check TFT_eSPI configuration"
>       ^~~~~
> /Users/xxx/ESPprojects/hello_world/components/TFT_eSPI/TFT_config.h:140:6: error: #error "Invalid Reset pin. Check TFT_eSPI configuration"
>      #error "Invalid Reset pin. Check TFT_eSPI configuration"
>       ^~~~~
> /Users/xxx/ESPprojects/hello_world/components/TFT_eSPI/TFT_config.h:148:10: error: #error "Invalid backlight control pin. Check TFT_eSPI configuration"
>          #error "Invalid backlight control pin. Check TFT_eSPI configuration"
>           ^~~~~
> /Users/xxx/ESPprojects/hello_world/components/TFT_eSPI/TFT_config.h:231:10: error: #error "Invalid MOSI pin. Check TFT_eSPI configuration"
>          #error "Invalid MOSI pin. Check TFT_eSPI configuration"
>           ^~~~~
> /Users/xxx/ESPprojects/hello_world/components/TFT_eSPI/TFT_config.h:237:10: error: #error "Invalid Clock pin. Check TFT_eSPI configuration"
>          #error "Invalid Clock pin. Check TFT_eSPI configuration"
>           ^~~~~

I noticed there is a KConfig file, which sets up menuconfig items.... and a bunch of menuconfig items appeared:
<img width="427" alt="Screenshot 2022-11-05 at 11 27 33 AM" src="https://user-images.githubusercontent.com/8603169/200135357-2996b954-655c-4b90-948b-28c44e687421.png">
These options are also listed in sdkconifg file.  

I have a feeling these menuconfig settings are conflicting/overwriting the setting in User_Setup_select.h.

If we're using esp-idf, how should we set the settings for use with our display?

Thank you.

**Comment by goutamreddy** [comment]
Okay, I went through and set all the parameters (as best as I understood) in menuconfig to match those in the file User_Setups/Setup25_TTGO_T_Display.h

![Screenshot 2022-11-05 at 2 11 01 PM](https://user-images.githubusercontent.com/8603169/200141460-25ae1239-c852-4469-b66f-e345083ae6b3.png)
![Screenshot 2022-11-05 at 2 11 27 PM](https://user-images.githubusercontent.com/8603169/200141476-4028843e-26d9-41cd-b9e9-303b8cc543a2.png)
![Screenshot 2022-11-05 at 2 11 50 PM](https://user-images.githubusercontent.com/8603169/200141486-44bf6ecf-824b-4633-a9b6-7404bbe2e224.png)


`// Setup for the TTGO T Display
#define USER_SETUP_ID 25

// See SetupX_Template.h for all options available

#define ST7789_DRIVER
#define TFT_SDA_READ   // Display has a bidirectional SDA pin

#define TFT_WIDTH  135
#define TFT_HEIGHT 240

#define CGRAM_OFFSET      // Library will add offsets required

//#define TFT_MISO -1

#define TFT_MOSI            19
#define TFT_SCLK            18
#define TFT_CS              5
#define TFT_DC              16
#define TFT_RST             23

#define TOUCH_CS            -1

#define TFT_BL          4  // Display backlight control pin

#define TFT_BACKLIGHT_ON HIGH  // HIGH or LOW are options

#define LOAD_GLCD
#define LOAD_FONT2
#define LOAD_FONT4
#define LOAD_FONT6
#define LOAD_FONT7
#define LOAD_FONT8
#define LOAD_GFXFF

#define SMOOTH_FONT

//#define SPI_FREQUENCY  27000000
  #define SPI_FREQUENCY  40000000


#define SPI_READ_FREQUENCY  6000000 // 6 MHz is the maximum SPI read speed for the ST7789V`



This seems a bit hackish (to have to redefine everything in menuconfig), so hopefully there is a better way to setup the configuration for esp-idf.

**Comment by Bodmer** [comment]
The support was added by a user. Have a look at #1770

**Comment by goutamreddy** [comment]
Thanks for finding that.  I understand what happened.

My assesment is that is your code is super dead-simple to use.... just uncomment the specific board you're using, and everything is taken care of re: pins, settings, etc...  The user doesn't really need to think through / lookup any of those settings.

With the addition of this KConfig file, the user can now set the options from menuconfig.  However, this overrides whatever settings are set in User_Setup_Select.h.

My understanding now is that if you install the TFT_eSPI as a component in esp-idf... it will automatically create the menuconfig options from the KConfig file... and a user IS FORCED TO SET all of those settings themselves in the sdkconfig file.... (e.g. by using menuconfig).

As in, I had to lookup that the T-Display board was board #25, and then find the appropriate settings in the .h files, and then enter them into manually into menuconfig...     This takes away from the "seamlessness" of the TFT_eSPI library, but exposes the user to more direct control -  I suppose it is a tradeoff-  but at the moment undocumented.

My only suggestion would be to indicate to users that to use TFT_eSPI as a component of esp-idf, they can either:
1) manually lookup and enter those settings in menuconfig
OR
2) delete the KConfig file and run idf.py reconfigure after tweaking the User_Setup_Select.h file (per the normal/usual method).

also- thanks again for building such an easy-to-use library!

**Reply by Bodmer** [reply]
Instructions added here:
https://github.com/Bodmer/TFT_eSPI/tree/master/docs/ESP-IDF

**Reply by goutamreddy** [reply]
@Bodmer  added a pull request that clarified the instruction file you created-

thanks again for making this library!

**Reply by electronicsguy** [reply]
Isn't this still using Arduino as a component? Its not pure esp-idf per se, is it?

**Reply by mayur9095** [reply]
yes, its using arduino as cmponent and its kind of hack to make it work on esp-idf.
